### PR TITLE
Add node model

### PR DESCRIPTION
**Pull request by stefanprobst**
This is for #11480

Questions:
* should nodeModel methods be called with `context`, or with `{ path: context: path }`?
* `getNodesByType` and `getNodeByType` handle GraphQL abstract types, what about `runQuery`?
* what else to put on nodeModel?
* `runQuery` signature (e.g. GraphQL type or typeName as arg)?
* why do we have `TypeConflictReporter` (instead of logging conflict right away)?

### CODE REVIEW

**Comment by freiksenet** [comment]
> why do we have `TypeConflictReporter` (instead of logging conflict right away)?

I think so we can log them all at the end in one place.

File: packages/gatsby/src/db/__tests__/node-tracking-test.js [hunk 1]
```diff
@@ -89,7 +88,19 @@ describe(`track root nodes`, () => {
     let type
 
     beforeAll(async () => {
-      type = (await nodeTypes.buildAll({})).testNode.nodeObjectType
+      const { SchemaComposer } = require(`graphql-compose`)
```
Comment (freiksenet) [inline packages/gatsby/src/db/__tests__/node-tracking-test.js]: Considering this is a such common pattern, is our API a bit wrong? Ideally we would be able to call `addInferredType({ schemaComposer, typeName: 'Test', nodeStore: fakeNodeStore, })`, what is preventing that from happening?

File: packages/gatsby/src/schema/types/NodeInterface.js [hunk 2]
```diff
@@ -22,13 +21,19 @@ const getOrCreateNodeInterface = schemaComposer => {
...
+          return context.nodeModel.getNode(source.parent, { path })
+        },
       },
       children: {
         type: `[Node]!`,
-        resolve: async (source, args, context, info) =>
-          source.children.map(context.nodeModel.getNode),
+        resolve: (source, args, context, info) => {
+          const { path } = context
+          return source.children.map(id =>
```
Comment (freiksenet) [inline packages/gatsby/src/schema/types/NodeInterface.js]: The fact that we use this pattern several times, means we need `getNodes` as a model method.

File: packages/gatsby/src/schema/node-model.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  return result
+}
+
+const nodeModel = {
+  findRootNodeAncestor,
+  getNode: withPageDependencies(getNodeById),
+  getNodeByType: withPageDependencies(getNodeByGQLTypeName),
+  getNodes: withPageDependencies(getNodes),
+  getNodesByType: withPageDependencies(getNodesByGQLTypeName),
+  // TODO: What should this return? Schema types? Types in the store?
```
Comment (freiksenet) [inline packages/gatsby/src/schema/node-model.js]: We can't really return schema types, so it should return type names.

File: packages/gatsby/src/schema/node-model.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  // hasNodeChanged,
+  // getNodeAndSavePathDependency,
+  // TODO: Rethink the signature of runQuery!
+  runQuery,
+} = require(`../db/nodes`)
+const { findRootNodeAncestor } = require(`../db/node-tracking`)
+const createPageDependency = require(`../redux/actions/add-page-dependency`)
+const { store } = require(`../redux`)
+
+const withPageDependencies = fn => async (args, pageDependencies) => {
```
Comment (freiksenet) [inline packages/gatsby/src/schema/node-model.js]: Really cool.

File: packages/gatsby/src/schema/schema.js [hunk 5]
```diff
@@ -285,10 +272,14 @@ function createChildrenField(typeName) {
...
-          .map(context.nodeModel.getNode)
-          .filter(nodeIsOfType(typeName))
+      async resolve(source, args, context) {
+        const { path } = context
+        const result = await Promise.all(
+          source.children.map(id =>
+            context.nodeModel.getNodeByType({ id, type: typeName }, { path })
+          )
+        )
+        return result.filter(Boolean)
```
Comment (freiksenet) [inline packages/gatsby/src/schema/schema.js]: Why can it have non-true values? getNodeByType with typename should filter the stuff.

File: packages/gatsby/src/schema/schema.js [hunk 6]
```diff
@@ -298,10 +289,14 @@ function createChildField(typeName) {
...
-          .map(context.nodeModel.getNode)
-          .find(nodeIsOfType(typeName))
+      async resolve(source, args, context) {
+        const { path } = context
+        const result = await Promise.all(
+          source.children.map(id =>
+            context.nodeModel.getNodeByType({ id, type: typeName }, { path })
+          )
+        )
+        return result.find(Boolean)
```
Comment (freiksenet) [inline packages/gatsby/src/schema/schema.js]: Ditto as above.

File: packages/gatsby/src/schema/types/Date.js [hunk 7]
```diff
@@ -61,7 +61,34 @@ export function isDate(value) {
...
+  } else if (difference) {
+    return moment().diff(
+      moment.utc(normalizedDate, ISO_8601_FORMAT, true).locale(locale),
+      difference
+    )
+  }
+  return normalizedDate
+}
+
+export const dateResolver = {
```
Comment (freiksenet) [inline packages/gatsby/src/schema/types/Date.js]: :+1:

**Comment by freiksenet** [comment]
> `getNodesByType` and `getNodeByType` handle GraphQL abstract types, what about `runQuery`?
> * `runQuery` signature (e.g. GraphQL type or typeName as arg)?

Probably should have same behaviour. So probably graphql type.

> * should nodeModel methods be called with `context`, or with `{ path: context: path }`?

Let's have all arguments explicit, so we don't pass context or info objects directly to model layer.

>  * what else to put on nodeModel?

getNodes that gets multiple nodes by ids.

**Comment by freiksenet** [comment]
Really happy about this stuff in general, great work. :+1:

File: packages/gatsby/src/db/__tests__/node-tracking-test.js (hunk 1 above)
Comment (stefanprobst) [reply packages/gatsby/src/db/__tests__/node-tracking-test.js]: Oh this would work perfectly, if we provide a nodeStore

**Pull request merged**